conjure-http/src/server.rs

```diff
@@ -300,3 +327,334 @@ impl<'a> RequestContext<'a> {
+pub trait AsyncSerializeResponse<T, W> {
+    /// Serializes the response.
+    fn serialize(
+        request_headers: &HeaderMap,
+        value: T,
+    ) -> Result<Response<AsyncResponseBody<W>>, Error>;
+}
+
+/// A serializer which encodes `()` as an empty body and status code of `204 No Content`.
+pub enum EmptyResponseSerializer {}
+
+impl EmptyResponseSerializer {
+    fn serialize_inner<T>(body: T) -> Result<Response<T>, Error> {
+        let mut response = Response::new(body);
+        *response.status_mut() = StatusCode::NO_CONTENT;
+        Ok(response)
+    }
+}
+
+impl<W> SerializeResponse<(), W> for EmptyResponseSerializer {
+    fn serialize(_: &HeaderMap, _: ()) -> Result<Response<ResponseBody<W>>, Error> {
+        Self::serialize_inner(ResponseBody::Empty)
+    }
+}
+
+impl<W> AsyncSerializeResponse<(), W> for EmptyResponseSerializer {
+    fn serialize(_: &HeaderMap, _: ()) -> Result<Response<AsyncResponseBody<W>>, Error> {
+        Self::serialize_inner(AsyncResponseBody::Empty)
+    }
+}
+
+/// A serializer which acts like a Conjure-generated client would.
+pub enum ConjureResponseSerializer {}
+
+impl ConjureResponseSerializer {
+    fn serialize_inner<T, B>(
+        value: T,
+        make_body: impl FnOnce(Bytes) -> B,
+    ) -> Result<Response<B>, Error>
+    where
+        T: Serialize,
+    {
+        let body = json::to_vec(&value).map_err(Error::internal)?;
+
+        let mut response = Response::new(make_body(body.into()));
+        response
+            .headers_mut()
+            .insert(CONTENT_TYPE, APPLICATION_JSON);
+
+        Ok(response)
+    }
+}
+
+impl<T, W> SerializeResponse<T, W> for ConjureResponseSerializer
+where
+    T: Serialize,
+{
+    fn serialize(
+        _request_headers: &HeaderMap,
+        value: T,
+    ) -> Result<Response<ResponseBody<W>>, Error> {
+        Self::serialize_inner(value, ResponseBody::Fixed)
+    }
+}
+
+impl<T, W> AsyncSerializeResponse<T, W> for ConjureResponseSerializer
+where
+    T: Serialize,
+{
+    fn serialize(
+        _request_headers: &HeaderMap,
+        value: T,
+    ) -> Result<Response<AsyncResponseBody<W>>, Error> {
+        Self::serialize_inner(value, AsyncResponseBody::Fixed)
+    }
+}
+
+/// A trait implemented by header decoders used by custom Conjure server trait implementations.
+pub trait DecodeHeader<T> {
+    /// Decodes the value from headers.
+    fn decode<'a, I>(headers: I) -> Result<T, Error>
+    where
+        I: IntoIterator<Item = &'a HeaderValue>;
+}
+
+/// A trait implemented by URL parameter decoders used by custom Conjure server trait
+/// implementations.
+pub trait DecodeParams<T> {
+    /// Decodes the value from the sequence of values.
+    ///
+    /// The values have already been percent-decoded.
+    fn decode<I>(params: I) -> Result<T, Error>
+    where
+        I: IntoIterator,
+        I::Item: AsRef<str>;
+}
+
+/// A decoder which converts a single value using its [`FromStr`] implementation.
+pub enum FromStrDecoder {}
```

Comment:
On the client side we have separate types for header and params decoders, but that doesn't seem necessary so I've combined them here. If that seems right we can do some renames and deprecations in the client side.